File: js/midi-file.js
// ===========================================================================
// Minimal MIDI file parser — no external dependencies
// ===========================================================================

/**
 * Parse a MIDI file ArrayBuffer.
 * Returns { ppq, events: [{tick, midiNote, velocity}] }
 */
export function parseMidiFile(arrayBuffer){
  const data = new DataView(arrayBuffer);
  const bytes = new Uint8Array(arrayBuffer);
  let pos = 0;

  function readUint32(){ const v = data.getUint32(pos, false); pos += 4; return v; }
  function readUint16(){ const v = data.getUint16(pos, false); pos += 2; return v; }
  function readByte(){ return bytes[pos++]; }

  function readVLQ(){
    let value = 0;
    let b;
    do {
      b = readByte();
      value = (value << 7) | (b & 0x7F);
    } while (b & 0x80);
    return value;
  }

  // Header chunk
  const headerMagic = readUint32();
  if (headerMagic !== 0x4D546864) throw new Error('Not a MIDI file');
  const headerLen = readUint32(); // should be 6
  const format = readUint16();
  const ntrks = readUint16();
  const timeDivision = readUint16();

  if (timeDivision & 0x8000) throw new Error('SMPTE time division not supported');
  const ppq = timeDivision;

  const allEvents = [];

  for (let t = 0; t < ntrks; t++){
    const trackMagic = readUint32();
    if (trackMagic !== 0x4D54726B) {
      // skip unknown chunk
      const len = readUint32();
      pos += len;
      continue;
    }
    const trackLen = readUint32();
    const trackEnd = pos + trackLen;

    let tick = 0;
    let runningStatus = 0;

    while (pos < trackEnd){
      const delta = readVLQ();
      tick += delta;

      let statusByte = bytes[pos];

      if (statusByte & 0x80){
        // real status byte
        runningStatus = statusByte;
        pos++;
      } else {
        // running status — use last status
        statusByte = runningStatus;
        // don't advance pos, the data byte is next
      }

      const type = statusByte & 0xF0;
      const channel = statusByte & 0x0F;

      if (type === 0xFF){
        // meta event
        const metaType = readByte();
        const metaLen = readVLQ();
        pos += metaLen;
        runningStatus = 0; // meta events reset running status
      } else if (statusByte === 0xF0 || statusByte === 0xF7){
        // sysex
        const sysexLen = readVLQ();
        pos += sysexLen;
        runningStatus = 0;
      } else if (type === 0x90){
        // note on
        const note = readByte();
        const vel = readByte();
        if (vel > 0){
          allEvents.push({ tick, midiNote: note, velocity: vel });
        }
        // note on with vel=0 treated as note off — ignore for drums
      } else if (type === 0x80){
        // note off — skip
        pos += 2;
      } else if (type === 0xA0){
        // aftertouch
        pos += 2;
      } else if (type === 0xB0){
        // control change
        pos += 2;
      } else if (type === 0xC0){
        // program change
        pos += 1;
      } else if (type === 0xD0){
        // channel pressure
        pos += 1;
      } else if (type === 0xE0){
        // pitch bend
        pos += 2;
      } else {
        // unknown — try to skip safely by breaking
        break;
      }
    }

    pos = trackEnd;
  }

  allEvents.sort((a, b) => a.tick - b.tick);
  return { ppq, events: allEvents };
}

/**
 * GM drum map: map MIDI note to lane index 0-5
 * 0=crash, 1=hihat, 2=snare, 3=kick, 4=tom, 5=ride
 */
export function midiNoteToLane(note){
  // kick
  if (note === 35 || note === 36) return 3;
  // snare
  if (note === 38 || note === 40 || note === 37 || note === 39) return 2;
  // hihat
  if (note === 42 || note === 44 || note === 46) return 1;
  // toms
  if (note === 41 || note === 43 || note === 45 || note === 47 || note === 48 || note === 50) return 4;
  // crash
  if (note === 49 || note === 52 || note === 55 || note === 57) return 0;
  // ride
  if (note === 51 || note === 53 || note === 56 || note === 59) return 5;
  return undefined;
}

/**
 * Export a pattern as a MIDI file ArrayBuffer (Type 0, channel 10, GM drums).
 * pattern: array of {lane, steps[]} where steps is boolean[]
 * stepsPerBeat: 2 (eighth notes per quarter note)
 * ppq: 96
 */
export function exportMidiFile(lanePatterns, steps, bpm){
  const ppq = 96;
  // 8th note = 1 step = ppq/2 ticks
  const ticksPerStep = ppq / 2;
  const microsecondsPerBeat = Math.round(60000000 / bpm);

  // GM note numbers per lane: crash=49, hihat=42, snare=38, kick=36, tom=45, ride=51
  const GM_NOTES = [49, 42, 38, 36, 45, 51];

  // Collect all events [{tick, note, type}] type: 'on'|'off'
  const rawEvents = [];

  lanePatterns.forEach((pattern, laneIdx) => {
    const note = GM_NOTES[laneIdx];
    pattern.forEach((on, stepIdx) => {
      if (on){
        const tick = stepIdx * ticksPerStep;
        rawEvents.push({ tick, note, type: 'on', vel: 100 });
        rawEvents.push({ tick: tick + Math.floor(ticksPerStep * 0.5), note, type: 'off', vel: 0 });
      }
    });
  });

  rawEvents.sort((a, b) => a.tick - b.tick || (a.type === 'off' ? -1 : 1));

  // Encode variable-length quantity
  function encodeVLQ(value){
    if (value === 0) return [0];
    const bytes = [];
    bytes.unshift(value & 0x7F);
    value >>= 7;
    while (value > 0){
      bytes.unshift((value & 0x7F) | 0x80);
      value >>= 7;
    }
    return bytes;
  }

  // Build track data
  const trackBytes = [];

  // Tempo meta event: FF 51 03 tt tt tt
  trackBytes.push(...encodeVLQ(0)); // delta 0
  trackBytes.push(0xFF, 0x51, 0x03);
  trackBytes.push((microsecondsPerBeat >> 16) & 0xFF);
  trackBytes.push((microsecondsPerBeat >> 8) & 0xFF);
  trackBytes.push(microsecondsPerBeat & 0xFF);

  // Note events
  let currentTick = 0;
  rawEvents.forEach(ev => {
    const delta = ev.tick - currentTick;
    currentTick = ev.tick;
    trackBytes.push(...encodeVLQ(delta));
    // Channel 10 (index 9): 0x99 = note-on ch10, 0x89 = note-off ch10
    if (ev.type === 'on'){
      trackBytes.push(0x99, ev.note, ev.vel);
    } else {
      trackBytes.push(0x89, ev.note, ev.vel);
    }
  });

  // End of track: FF 2F 00
  trackBytes.push(...encodeVLQ(0));
  trackBytes.push(0xFF, 0x2F, 0x00);

  // Assemble file
  const totalLen = 14 + 8 + trackBytes.length;
  const buf = new ArrayBuffer(totalLen);
  const view = new DataView(buf);
  let offset = 0;

  function writeUint32(v){ view.setUint32(offset, v, false); offset += 4; }
  function writeUint16(v){ view.setUint16(offset, v, false); offset += 2; }
  function writeBytes(arr){ arr.forEach(b => { view.setUint8(offset, b); offset++; }); }

  // Header
  writeUint32(0x4D546864); // 'MThd'
  writeUint32(6);          // header length
  writeUint16(0);          // format 0
  writeUint16(1);          // 1 track
  writeUint16(ppq);        // ppq

  // Track chunk
  writeUint32(0x4D54726B); // 'MTrk'
  writeUint32(trackBytes.length);
  writeBytes(trackBytes);

  return buf;
}

File: ios/SP808Killa/Domain/MIDIFile.swift
import Foundation

enum MIDIFileError: Error, Equatable {
    case notMIDI
    case smpteUnsupported
    case truncated
}

struct MIDINoteEvent: Equatable {
    let tick: Int
    let note: Int
    let velocity: Int
}

/// Minimal SMF parser — port of `parseMidiFile` in `js/midi-file.js`. Handles
/// header/track chunks, PPQ timing, running status, note-on (vel>0); skips meta/
/// sysex/other channel messages; rejects SMPTE division.
enum MIDIFileParser {
    static func parse(_ data: Data) throws -> (ppq: Int, events: [MIDINoteEvent]) {
        let b = [UInt8](data)
        var pos = 0

        func need(_ n: Int) throws { if pos + n > b.count { throw MIDIFileError.truncated } }
        func u32() throws -> Int {
            try need(4)
            defer { pos += 4 }
            return Int(b[pos]) << 24 | Int(b[pos+1]) << 16 | Int(b[pos+2]) << 8 | Int(b[pos+3])
        }
        func u16() throws -> Int {
            try need(2)
            defer { pos += 2 }
            return Int(b[pos]) << 8 | Int(b[pos+1])
        }
        func byte() throws -> Int { try need(1); defer { pos += 1 }; return Int(b[pos]) }
        func vlq() throws -> Int {
            var value = 0, c: Int
            repeat { c = try byte(); value = (value << 7) | (c & 0x7F) } while c & 0x80 != 0
            return value
        }

        guard try u32() == 0x4D546864 else { throw MIDIFileError.notMIDI }   // 'MThd'
        _ = try u32()              // header length (6)
        _ = try u16()              // format
        let ntrks = try u16()
        let timeDivision = try u16()
        if timeDivision & 0x8000 != 0 { throw MIDIFileError.smpteUnsupported }
        let ppq = timeDivision

        var events: [MIDINoteEvent] = []

        for _ in 0..<ntrks {
            let magic = try u32()
            let len = try u32()
            if magic != 0x4D54726B { pos += len; continue }   // skip non-'MTrk'
            let end = pos + len
            var tick = 0
            var running = 0

            while pos < end {
                tick += try vlq()
                try need(1)
                var status = Int(b[pos])
                if status & 0x80 != 0 { running = status; pos += 1 } else { status = running }
                let type = status & 0xF0

                if status == 0xFF {
                    _ = try byte()            // meta type
                    let mlen = try vlq()
                    pos += mlen
                    running = 0
                } else if status == 0xF0 || status == 0xF7 {
                    let slen = try vlq()
                    pos += slen
                    running = 0
                } else if type == 0x90 {
                    let note = try byte(), vel = try byte()
                    if vel > 0 { events.append(MIDINoteEvent(tick: tick, note: note, velocity: vel)) }
                } else if type == 0x80 || type == 0xA0 || type == 0xB0 || type == 0xE0 {
                    pos += 2
                } else if type == 0xC0 || type == 0xD0 {
                    pos += 1
                } else {
                    break
                }
            }
            pos = end
        }

        events.sort { $0.tick < $1.tick }
        return (ppq, events)
    }

    /// Build a custom `Lesson` from parsed events (port of `lessonFromMidiEvents`).
    /// Quantizes to eighth-note (ppq/2) steps; one full-length pattern per lane.
    static func lesson(name: String, events: [MIDINoteEvent], ppq: Int) -> Lesson {
        let halfBeat = max(1, ppq / 2)
        var beats: [(lane: Int, beat: Int)] = events.compactMap {
            guard let lane = GMDrumMapper.lane(forNote: $0.note) else { return nil }
            return (lane.rawValue, Int((Double($0.tick) / Double(halfBeat)).rounded()))
        }
        beats.sort { $0.beat != $1.beat ? $0.beat < $1.beat : $0.lane < $1.lane }
        let lastBeat = beats.map(\.beat).max() ?? 0
        let totalBeats = max(16, lastBeat + 1)

        var patterns: [PatternLine] = []
        for lane in DrumLane.allCases {
            let hits = Set(beats.filter { $0.lane == lane.rawValue }.map(\.beat))
            guard !hits.isEmpty else { continue }
            let pattern = (0..<totalBeats).map { hits.contains($0) ? "x" : "." }.joined()
            patterns.append(PatternLine(lane: lane.key, pattern: pattern))
        }
        if patterns.isEmpty {
            patterns = [PatternLine(lane: "kick", pattern: String(repeating: ".", count: totalBeats))]
        }
        return Lesson(name: name, bpm: 80, tip: "MIDI file: \(name)",
                      difficulty: "Custom", genre: "Import", patterns: patterns)
    }
}

/// SMF Type-0 exporter — port of `exportMidiFile`. Channel 10, PPQ 96, GM drum
/// notes, tempo meta. Byte-compatible with the web export.
enum MIDIFileExporter {
    static func export(lanePatterns: [[Bool]], bpm: Int) -> Data {
        let ppq = 96
        let ticksPerStep = ppq / 2
        let usPerBeat = Int((60_000_000.0 / Double(bpm)).rounded())

        struct Raw { let tick: Int; let note: Int; let isOn: Bool; let vel: Int }
        var raw: [Raw] = []
        for (laneIdx, pattern) in lanePatterns.enumerated() {
            guard laneIdx < GMDrumMapper.gmNotes.count else { break }
            let note = GMDrumMapper.gmNotes[laneIdx]
            for (step, on) in pattern.enumerated() where on {
                let tick = step * ticksPerStep
                raw.append(Raw(tick: tick, note: note, isOn: true, vel: 100))
                raw.append(Raw(tick: tick + ticksPerStep / 2, note: note, isOn: false, vel: 0))
            }
        }
        // tick asc; at same tick offs before ons (matches the web sort).
        raw.sort { $0.tick != $1.tick ? $0.tick < $1.tick : (!$0.isOn && $1.isOn ? true : false) }

        func vlq(_ value: Int) -> [UInt8] {
            if value == 0 { return [0] }
            var v = value
            var out: [UInt8] = [UInt8(v & 0x7F)]
            v >>= 7
            while v > 0 { out.insert(UInt8((v & 0x7F) | 0x80), at: 0); v >>= 7 }
            return out
        }

        var track: [UInt8] = []
        track += vlq(0)
        track += [0xFF, 0x51, 0x03,
                  UInt8((usPerBeat >> 16) & 0xFF), UInt8((usPerBeat >> 8) & 0xFF), UInt8(usPerBeat & 0xFF)]
        var cur = 0
        for ev in raw {
            track += vlq(ev.tick - cur)
            cur = ev.tick
            track += [ev.isOn ? 0x99 : 0x89, UInt8(ev.note), UInt8(ev.vel)]
        }
        track += vlq(0)
        track += [0xFF, 0x2F, 0x00]

        var out: [UInt8] = []
        func u32(_ v: Int) { out += [UInt8((v >> 24) & 0xFF), UInt8((v >> 16) & 0xFF), UInt8((v >> 8) & 0xFF), UInt8(v & 0xFF)] }
        func u16(_ v: Int) { out += [UInt8((v >> 8) & 0xFF), UInt8(v & 0xFF)] }
        u32(0x4D546864); u32(6); u16(0); u16(1); u16(ppq)         // MThd, fmt0, 1 track, ppq
        u32(0x4D54726B); u32(track.count); out += track          // MTrk
        return Data(out)
    }
}
